lib/linter/linter.js

```diff
@@ -1119,6 +1119,10 @@ function runRules(sourceCode, configuredRules, ruleMapper, parserName, languageO
             };
         }
 
+        if (typeof ruleListeners !== "object" || Array.isArray(ruleListeners) || ruleListeners === null) {
+            throw new Error(`The create() function for rule ${ruleId} did not return an object.`);
+        }
```

Comment:
Does ESLint currently crash if `ruleListeners` is an array? If it currently crashes only on `undefined` and `null`, maybe we should cover only them.

Reply:
I agree. We shouldn’t be adding cases that will fail if they don’t already.

Reply:
Would you folks support me opening an issue to add a breaking change to ESLint v9 to only allow rules to return objects? I suspect it's unintentional that we allow them to return arrays or random literal values right now. This seems small enough to not require an RFC.

Reply:
> Would you folks support me opening an issue to add a breaking change to ESLint v9 to only allow rules to return objects? I suspect it's unintentional that we allow them to return arrays or random literal values right now. This seems small enough to not require an RFC.

Makes sense to me. We could continue the discussion in the issue and decide there whether an RFC would be necessary.

Reply:
I’d say go for it. We can definitely discuss for v9. 

As a note: I’m not sure there’s anything wrong with returning an array. Arrays can have other properties which could be visitors. Chances are people won’t mean to do that, but ESLint would work ok.

Reply:
Opened issue: https://github.com/eslint/eslint/issues/16086